lib/board.rb

```
@@ -15,9 +15,37 @@ def mark_board(coordinates)
     grid[coordinates[:y] - 1][coordinates[:x] - 1] = 'X'
   end
 
-
   def validate_choice(coordinates)
-    grid[coordinates[:y] - 1][coordinates[:x] - 1] == 'X' ? false : true
+    valid = { 'bool' => false, 'error' => nil }
+    if coords_0?(coordinates)
+      valid['error'] = 1
+    elsif is_a_num?(coordinates) && already_marked?(coordinates)
+      valid['error'] = 2
+    else
+      valid['bool'] = true
+    end
+    valid
+  end
+
```

I think you could start looking into raising custom exceptions here - that way you could raise an error for 'already marked' and raise a different error for 'invalid input' or whatever.

You could then handle those errors further up and display the messages accordingly